ginivision/src/doc/source/guide-for-open-with.rst

```diff
@@ -0,0 +1,182 @@
+How to Upload PDFs and Images
+=============================
+
+General considerations
+----------------------
+
+Enabling your app to open PDFs and images allows your users to open any kind of files which are identified by the OS as PDFs or images. Make sure to check the MIME type and that the file size is below an acceptable threshold (5MB for example). It is also advisable to check the first bytes of the incoming files to determine their type and allow only PDFs and known image types.
+
+Registering PDF and image file types
+------------------------------------
+
+Add the following intent filter to the *Activity* in your *AndroidManifest.xml* you wish to receive incoming PDFs and images:
+
+.. code-block:: xml
+
+    <activity android:name=".ui.MyActivity">
+        <intent-filter android:label="@string/app_name">
+            <action android:name="android.intent.action.VIEW" />
+            <action android:name="android.intent.action.SEND" />
+            <category android:name="android.intent.category.DEFAULT" />
+            <data android:mimeType="image/*" />
+            <data android:mimeType="application/pdf" />
+        </intent-filter>
+    </activity>
+
+.. note::
+
+    We recommend adding `ACTION_VIEW <https://developer.android.com/reference/android/content/Intent.html#ACTION_VIEW>`_ to the intent filter to also allow users to send PDFs and images to your app from apps which don’t implement sharing with `ACTION_SEND <https://developer.android.com/reference/android/content/Intent.html#ACTION_SEND>`_ but enable viewing the PDF or file with other apps.
+
+Documentation
+^^^^^^^^^^^^^
+
+- https://developer.android.com/training/basics/intents/filters.html
+
+Reading files from external storage
+-----------------------------------
+
+Since Android 4.4 (API Level 19) the `READ_EXTERNAL_STORAGE <https://developer.android.com/reference/android/Manifest.permission.html#READ_EXTERNAL_STORAGE>`_ permission has to be declared to access files outside of the app-specific directories. Declare this permission in your *AndroidManifest.xml*:
+
+.. code-block:: xml
+
+    <uses-permission android:name="android.permission.READ_EXTERNAL_STORAGE" />
+ 
+Since the introduction of `Run Time Permissions <https://developer.android.com/training/permissions/requesting.html>`_ with Android 6.0 (API Level 23) your app also has to request permission from the user before accessing a PDF or an image file outside of the app-specific directories. 
+ 
+We recommend checking for the permission whenever your app receives a PDF or an image and before accessing it. If an explanation is requested by the system to be shown to the user we highly recommend providing one.
+ 
+Handling the case where permission is not granted is required.  We recommend showing an explanation and providing a link to your app’s *App info* page in the *Settings* app with a short guide on how to grant the permission there. 
```

Comment:
Suggested wording for first sentence: "It is required to handle the case where the user did not grant permissions."